README
NAME
    Data::Header::Fields - encode and decode RFC822 header field lines

SYNOPSIS
            use IO::Any;
            my $email_msg = IO::Any->slurp([ 'path', 'to', 'email.eml' ]);
            my ($email_header, $email_body) = split(/^\s*$/m, $email_msg, 2);

            use Data::Header::Fields;
            my $dhf = Data::Header::Fields->new->decode(\$email_header);
            print 'From    - ', $dhf->get_value('From'), "\n";
            print 'Subject - ', $dhf->get_value('Subject'), "\n";
            print 'Date    - ', $dhf->get_value('Date'), "\n";
            print '--- cut ---', "\n";

            $dhf->set_field('To' => ' anyone@anywhere');
            $dhf->rm_fields('Received');
        
            print $dhf->encode();

WARNING
    experimental, use on your own risk :-)

DESCRIPTION
    RFC822 - Standard for ARPA Internet Text Messages
    (http://tools.ietf.org/html/rfc822#section-3.2) describes the format of
    header lines used in emails. The tricky part is the line folding.

    There are some "forks" of this standard. One of them is Debian
    RFC-2822-like fields and the other is RFC2425 that defines the so called
    vCard format. Data::Header::Fields is generic enough to serve as a base
    class to parse those as well.

    One of the main goals of the module is to be able to edit the headers
    while keeping the lines that were not changed untouched.

    For the moment this is all documentation. After more tests with vCards
    and using this module for the basic parsing in Parse::Deb::Control it
    will be stable enough.

    Currently this distribution is highly radioactive!

SEE ALSO
    http://tools.ietf.org/html/rfc2822 - Internet Message Format

AUTHOR
    Jozef Kutej



Changes
Revision history for Data-Header-Fields

0.04    27 Sep 2011
    - fix memory leaks in ::Line and ::Value
    - Data::v get_fields() with optional param name/value filter

0.03    08 Apr 2011
    - s/set_field/set_value/
    - s/update_fields/update_values/
    - s/set_field/set_value/
    - add Data::v::Param

0.02    01 Apr 2011
    - proper decoding of binary fields (photo, logo, sound, key)
    - every line value is an object

0.01    26 Mar 2011
    - initial


Build.PL
use strict;
use warnings;

use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Data::Header::Fields',
    license             => 'perl',
    dist_author         => 'Jozef Kutej <jkutej@cpan.org>',
    dist_version_from   => 'lib/Data/Header/Fields.pm',
    configure_requires => {
        'Module::Build' => '0.36',
    },
    requires => {
        'IO::Any'           => 0,
        'Carp'              => 0,
        'String::Escape'    => 0,
        'List::MoreUtils'   => 0,
        'Scalar::Util'      => 0,
        'MIME::QuotedPrint' => 0,
        'MIME::Base64'      => 0,    # Introduce make_path() and rm_tree() API extension.
        'Encode'            => 0,
    },
    build_requires => {
        'Test::More'        => 0,
        'Test::Deep'        => 0,
        'Test::Differences' => 0,
        'Test::Exception'   => 0,
    },
    add_to_cleanup      => [ 'Data-Header-Fields-*' ],
    create_readme       => 1,
    sign                => 1,
    meta_merge => {
        resources => {
            repository => 'git://github.com/jozef/Data-Header-Fields.git',
            bugtracker => 'http://github.com/jozef/Data-Header-Fields/issues',
        },
        keywords => [ qw/ rfc822 rfc2822 header parse edit / ],
    },
);
$builder->create_build_script();


lib/Data/v.pm
package Data::v;

=head1 NAME

=head1 SYNOPSIS

	use Data::v;

	my $vdata = Data::v->new->decode([ '..', 't', 'vcf', 'aldo.vcf' ]);
	my $vcard = $vdata->get_value('vcard');
	
	print 'version:   ', $vcard->get_value('version'), "\n";
	print 'full name: ', $vcard->get_value('fn'), "\n";
	print 'email:     ', $vcard->get_value('email'), "\n";
	
	my @cell_phones = $vcard->get_fields('tel');
	
	use List::MoreUtils 'any';
	my @cell_phones =
		map { $_->value->as_string }
		$vcard->get_fields('tel', 'type' => 'cell')
	;
	print 'cell:      ', join(', ', @cell_phones), "\n";
	
	print "\n";
	
	$vcard->set_value('email' => 'dada@internet');
	$vcard->rm_fields('rev', 'photo', 'adr', 'X-MS-OL-DEFAULT-POSTAL-ADDRESS', 'label');
	
	print $vdata->encode, "\n";

=head1 DESCRIPTION

=cut

use warnings;
use strict;

use Carp;
use Scalar::Util 'blessed';
use List::MoreUtils 'any';

our $VERSION = '0.03';

use base 'Data::Header::Fields';

our %v_types = (
	'VCARD' => 'Data::v::Card',
);

sub new {
	my $class = shift;
	return $class->SUPER::new(
		tight_folding => 1,
		key_cmp       => \&default_key_cmp,
		parent        => $class->_default_parent,
		@_,
	);
}

sub default_key_cmp {
	my $a = shift;
	my $b = shift;
	
	$a = lc $a;
	$b = lc $b;
	
	return 0
		if $b =~ qr{^$a (?: ; | $)}xms;
	
	return $a cmp $b;
}

sub decode {
	my $self = shift;
	my $any  = shift;
	
	my $return_self = ref $self;
	$self = $self->new()
		if not ref $self;

	my $dhf = $self->SUPER::decode($any);
	my $lines = $dhf->_lines;
	
	my @v_entries;
	my $v_type;
	while (my $line = shift @{$lines}) {
		if ($line->key eq 'BEGIN') {
			$v_type = $line->value->as_string;
			croak 'unknown v-type "'.$v_type.'"'
				if not $v_types{$v_type};
			
			my $v_data = $v_types{$v_type}->new();
			my $v_entry = ($v_types{$v_type}.'::Entry')->new(
				'key'           => $v_type,
				'value'         => $v_data,
				'parent'        => $self,
			);
			$v_data->parent($v_entry);
			push @v_entries, $v_entry;
			next;
		}
		elsif ($line->key eq 'END') {
			croak 'BEGIN and END mismatch "'.$v_type.'" ne "'.$line->value->as_string.'"'
				if $v_type ne $line->value->as_string;
			
			$v_type = undef;
			next;
		}
		
		push @{$v_entries[-1]->value->_lines}, $line;
	}
		
	foreach my $v_entry (@v_entries) {
		$v_entry->value->rebless_lines;
	}
	
	if (not $return_self) {
		return \@v_entries;
	}

	$dhf->_lines(\@v_entries);
	return $dhf;	
}

sub _read_lines {
	my $self = shift;
	my $any  = shift;
	
	my $fh = IO::Any->read($any);
	
	# put folded lines to an array http://tools.ietf.org/html/rfc2822#section-2.2.3
	my @lines;
	my $quoted_printable = 0;
	while (my $line = <$fh>) {
		# folded line
		if (($line =~ m/^\s/xms) or ($quoted_printable and ($lines[-1] =~ m/ = \r? \Z /xms))) {
			# ignore if the first line starts with white space
			next if not @lines;
			
			$lines[-1] .= $line;
			next;
		}
		
		# detect quoted-printable encoding which continues on a next lines when the line ends with "="
		my ($key, $value) = split(/:/, $line, 2);
		my @key_parts = split(/;/, $key);
		shift @key_parts;
		$quoted_printable = (any { $_ eq 'encoding=quoted-printable' } map { lc $_; } @key_parts);
				
		push @lines, $line;
	}
	
	close $fh;

	return @lines;	
}

sub _default_parent {
	return 'Data::Header::Fields';
}

sub parent {
	my $self   = shift;
	$self->{'parent'} = shift
		if @_;
	
	return (ref $self ? $self->{'parent'} : $self->_default_parent);
}

1;

package Data::v::Card;

use base 'Data::v';

use List::MoreUtils 'any';
use Carp 'croak';

sub version { return $_[0]->get_value('version') || '2.1'; }

sub rebless_lines {
	my $self = shift;
	
	foreach my $line (@{$self->_lines}) {
		$line = Data::v::Card::Line->new(
			line   => $line,
			parent => $self,
		);
	}
}

sub _default_parent {
	return 'Data::v::Card::Entry';
}

sub line_ending {
	my $self = shift;
	return $self->parent->parent->line_ending(@_);
}

sub get_fields {
	my $self        = shift;
	my $field_name  = shift or croak 'field_name argument is mandatory';
	my $param_name  = shift;
	my $param_value = shift;
	
	my @fields = $self->SUPER::get_fields($field_name);
	if (defined $param_name) {
		@fields = 
			grep { any { lc $_->value eq $param_value } $_->get_key_params($param_name) }
			@fields
		;
	}
	
	return @fields;
}

1;


package Data::v::Card::Line;

use base 'Data::Header::Fields::Line';

use Carp 'croak';
use MIME::QuotedPrint 'encode_qp', 'decode_qp';
use MIME::Base64 'encode_base64', 'decode_base64';
use Encode ();
use List::MoreUtils 'none';

use overload
	'""'  => \&as_string,
	'cmp' => \&Data::Header::Fields::Line::cmp,
;

sub new {
	my $class = shift;
	my $self  = $class->SUPER::new(
		@_,
	);
	$self->_decode_key_params;
	return $self;
}

sub version { return $_[0]->parent->version; }

sub params {
	my $self = shift;
	
	if (@_) {
		$self->{params} = shift;
	}
	$self->{params} = []
		if (not $self->{params});
	
	return $self->{params};
}

sub _decode_key_params {
	my $self = shift;
	
	my $key = $self->key;
	
	if ($key =~ m/^([^;]+);(.+)$/xms) {
		my $orig_key_name  = $1;
		my @raw_key_params = split /;/, $2;
		my @key_params;
		
		my $key_name = lc $orig_key_name;

		foreach my $key_param (@raw_key_params) {
			croak 'unknown key param "'.$key_param.'"'
				if $key_param !~ m/^ (?: ([^=]+) = )? (.+) $/xms;
			my $param_name = $1 || 'TYPE';
			my $param_str  = $2;
			
			push
				@key_params,
					map { Data::v::Param->new('name' => $param_name, 'value' => $_, 'parent' => $self) }
					(split(/,/, $param_str))
			;
		}
		
		$self->key($orig_key_name);
		$self->params(\@key_params);
		
		my $enc_type   = lc ($self->get_key_param_value('encoding') || '');
		if ($enc_type) {
			if ($enc_type eq 'quoted-printable') {
				$self->{value} = Data::Header::Fields::Value->new(
					decode_qp($self->{value})
				);
			}
			elsif ($enc_type eq 'base64') {
				$self->{value} = Data::Header::Fields::Value->new(
					decode_base64($self->{value})
				);
			}
			else {
				croak 'unknown encoding "'.$enc_type.'"';
			}
		}

		my $charset = lc ($self->get_key_param_value('charset') || '');
		$charset ||= 'utf8'
			if (none { $_ eq $key_name } qw(photo logo sound key));

		if ($charset) {
			$self->{'value'} = Data::Header::Fields::Value->new(
				eval { Encode::decode($charset, $self->{'value'}) }
			);
		}
	}
	
	if ((lc $self->key eq 'n') and (not $self->key->isa('Data::v::Card::Value::Name'))) {
		$self->{'value'} = Data::v::Card::Value::Name->new(    # not calling value() because the set doesn't affect the content of the value
			'value' => $self->value,
			'parent' => $self,
		);
	}
	elsif ((lc $self->key eq 'adr') and (not $self->key->isa('Data::v::Card::Value::Adr'))) {
		$self->{'value'} = Data::v::Card::Value::Adr->new(     # not calling value() because the set doesn't affect the content of the value
			'value'  => $self->value,
			'parent' => $self,
		);
	}
	
	return;
}

sub get_key_params {
	my $self       = shift;
	my $param_name = shift or croak 'param param_name is mandatory';
	my $params     = $self->params;
	
	$param_name = lc $param_name;
	return grep { lc $_->{'name'} eq $param_name } @{$params};
}

sub get_key_param {
	my $self       = shift;
	my $param_name = shift or croak 'param param_name is mandatory';
	
	my @params = $self->get_key_params($param_name);
	croak 'more then one key param with name "'.$param_name.'"'
		if @params > 1;
	
	return $params[0];	
}

sub get_key_param_value {
	my $self       = shift;
	my $param_name = shift or croak 'param param_name is mandatory';
	
	my $param = $self->get_key_param($param_name);
	return undef if not $param;
	return $param->{'value'};
}

sub update_key_params {
	my $self        = shift;
	my $param_name  = shift or croak 'param param_name is mandatory';
	my $param_value = shift;
	
	# updating via array set
	if (ref $param_value) {
		my @new_params     = @{$param_value};
		
		# update existing
		foreach my $param (@{$self->params}) {
			$param->value(shift @new_params)    # will returns undefs if depleeted
				if ($param->name eq $param_name);
		}
		
		# add any additional new
		foreach my $add_value (@new_params) {
			push @{$self->{params}}, Data::v::Param->new(
				'parent' => $self,
				'name'   => $param_name,
				'value'  => $add_value,
			);
		}
		
		# remove any additional old
		$self->{params} = [
			grep { defined $_->{'value'} }
			@{$self->{params}}
		];
		return $self;
	}

	my @params = (
		map {
			($_->{'name'} eq $param_name ? $_->{value} = $param_value : ());
			$_;
		} @{$self->params}
	);
	
	return $self;
}

sub set_key_param {
	my $self        = shift;
	my $param_name  = shift or croak 'param param_name is mandatory';
	my $param_value = shift;

	my @params = $self->get_key_params($param_name);
	if ((@params > 0) or (ref $param_value)) {
		$self->update_key_params($param_name, $param_value);
	}
	elsif (@params == 0) {
		push @{$self->params}, Data::v::Param->new('name' => $param_name, 'value' => $param_value, 'parent' => $self);
	}
	else { 
		croak 'more then one param field with name "'.$param_name.'"';
	}
	
	return $self;
}

sub rm_key_param {
	my $self        = shift;
	my $param_name  = shift or croak 'param param_name is mandatory';

	my @params = (
		grep {
			$_->name ne $param_name
		} @{$self->params}
	);
	$self->params(\@params);
	
	return $self;
}

sub _encode_key_params {
	my $self = shift;

	my $params = $self->params;
	return if scalar @{$params} == 0;
	my $key   = $self->key;

	my $charset = lc ($self->get_key_param_value('charset') || 'utf8');
	$self->{value} = eval { Encode::encode($charset, $self->{value}) };
	
	my $enc_type   = lc ($self->get_key_param_value('encoding') || '');
	if ($enc_type) {
		if ($enc_type eq 'quoted-printable') {
			$self->{value} = encode_qp($self->{value}, "");
		}
		elsif ($enc_type eq 'base64') {
			$self->{value} = encode_base64($self->{value}, "");
		}
		else {
			croak 'unknown encoding "'.$enc_type.'"';
		}
	}
	
	if ($self->version eq '2.1') {
		$key .= ';'.(
			join(
				';',
				(
					map  { $_->as_string }
					grep { defined $_->value }
					@{$params}
				),
			)
		);
	}
	elsif ($self->version ge '3.0') {
		my @types = map { $_->as_string } $self->get_key_params('type');
		$key .= ';'.(
			join(
				';',
				(
					map {
						(lc $_->name eq 'type')
						? ( @types ? ('TYPE='.join(',',splice(@types,0,scalar @types))) : () )
						: $_->as_string
					}
					grep { defined $_->value }
					@{$params}
				),
			)
		);
	}
	else {
		croak 'unsupported VCARD version '.$self->version;
	}
	
	$self->params(undef);
	$self->key($key);
	
	return;
}

sub as_string {
	my $self = shift;
	
	if (exists $self->{'original_line'}) {
		my $original_line = $self->{'original_line'};
		
		# make sure the line has line_ending, even the original one could be created without using ->new()
		$original_line .= $self->parent->line_ending
			if $original_line !~ m/ \n \Z /xms;
		
		return $original_line;
	}

	$self->_encode_key_params;

	my ($key, $value) = ($self->key, $self->value);
	#$value = String::Escape::printable($value);
	# FIXME ^^^ should be moved to _encode_key_params

	my $line = join(':', $key, $value);
	
	$line .= $self->parent->line_ending
		if $line !~ m/\n$/xms;

	$self->_decode_key_params;
	
	return $line;
}

1;

package Data::v::Card::Entry;

use base 'Data::Header::Fields::Line';

use overload
	'""'  => \&as_string,
	'cmp' => \&Data::Header::Fields::Line::cmp,
;

sub as_string {
	my $self = shift;

	return
		'BEGIN:VCARD'.$self->parent->line_ending
		.$self->value->as_string()
		.'END:VCARD'.$self->parent->line_ending
	;
}

1;

package Data::v::Card::Value::Name;

use base 'Data::Header::Fields::Value';

our @NAME_PART_TYPES = qw{family_name given_name additional_names honorific_prefixes honorific_suffixes};

use overload
	'""'  => \&Data::Header::Fields::Value::as_string,
	'cmp' => \&Data::Header::Fields::Value::cmp,
;

sub new {
	my $class = shift;
	my $self = $class->SUPER::new(@_);
	
	defined $self->{'value'}
	? $self->_parse_value()
	: $self->_update_value();
	
	return $self;
}

sub _update_value {
	my $self = shift;
	
	my @name_parts = map { $self->$_() } @NAME_PART_TYPES;
	# remove the undef fields from the end of the N
	while ((@name_parts) and (not defined $name_parts[-1])) {
		pop @name_parts;
	}
	@name_parts = map { defined $_ ? $_ : '' } @name_parts;
	
	$self->value(join(';', @name_parts));
	
	return $self;
}

sub _parse_value {
	my $self = shift;
	
	my $name_str = $self->{'value'};
	my @name_parts = split(/;/, $name_str);
	
	foreach my $name_part_type (@NAME_PART_TYPES) {
		my $name_part_value = shift @name_parts;
		$self->{$name_part_type} = $name_part_value;
	}
	
	return $self;
}

sub family_name {
	my $self   = shift;	
	
	if (@_) {
		$self->{'family_name'} = shift;
		$self->_update_value();
	}
	return $self->{'family_name'};
}
sub given_name {
	my $self   = shift;	
	if (@_) {
		$self->{'given_name'} = shift;
		$self->_update_value();
	}
	return $self->{'given_name'};
}
sub additional_names {
	my $self   = shift;	
	if (@_) {
		$self->{'additional_names'} = shift;
		$self->_update_value();
	}
	return $self->{'additional_names'};
}
sub honorific_prefixes {
	my $self   = shift;	
	if (@_) {
		$self->{'honorific_prefixes'} = shift;
		$self->_update_value();
	}
	return $self->{'honorific_prefixes'};
}
sub honorific_suffixes {
	my $self   = shift;	
	if (@_) {
		$self->{'honorific_suffixes'} = shift;
		$self->_update_value();
	}
	return $self->{'honorific_suffixes'};
}

1;

package Data::v::Card::Value::Adr;

use base 'Data::Header::Fields::Value';

our @ADR_PART_TYPES = qw{po_box ext_address street city state postal_code country};

use overload
	'""'  => \&Data::Header::Fields::Value::as_string,
	'cmp' => \&Data::Header::Fields::Value::cmp,
;

sub new {
	my $class = shift;
	my $self = $class->SUPER::new(@_);
	
	defined $self->{'value'}
	? $self->_parse_value()
	: $self->_update_value();
	
	return $self;
}

sub _update_value {
	my $self = shift;

	my @adr_parts = map { $self->$_() } @ADR_PART_TYPES;
	# remove the undef fields from the end of the N
	while ((@adr_parts) and (not defined $adr_parts[-1])) {
		pop @adr_parts;
	}
	@adr_parts = map { defined $_ ? $_ : '' } @adr_parts;
	
	$self->value(join(';', @adr_parts));
	
	return $self;
}

sub _parse_value {
	my $self = shift;
	
	my $adr_str = $self->{'value'};
	$adr_str =~ s/ \r? \n \Z//xms;
	my @adr_parts = split(/;/, $adr_str);
	
	foreach my $adr_part_type (@ADR_PART_TYPES) {
		my $adr_part_value = shift @adr_parts;		
		$self->{$adr_part_type} = $adr_part_value;
	}
	
	return $self;
}

sub po_box {
	my $self   = shift;

	if (@_) {
		$self->{'po_box'} = shift @_;
		$self->_update_value();
	}

	return $self->{'po_box'};
}
sub ext_address {
	my $self   = shift;	

	if (@_) {
		$self->{'ext_address'} = shift @_;
		$self->_update_value();
	}

	return $self->{'ext_address'};
}
sub street {
	my $self   = shift;	

	if (@_) {
		$self->{'street'} = shift @_;
		$self->_update_value();
	}

	return $self->{'street'};
}
sub city {
	my $self   = shift;	

	if (@_) {
		$self->{'city'} = shift @_;
		$self->_update_value();
	}

	return $self->{'city'};
}
sub state {
	my $self   = shift;	

	if (@_) {
		$self->{'state'} = shift @_;
		$self->_update_value();
	}

	return $self->{'state'};
}
sub postal_code {
	my $self   = shift;	

	if (@_) {
		$self->{'postal_code'} = shift @_;
		$self->_update_value();
	}

	return $self->{'postal_code'};
}
sub country {
	my $self   = shift;	

	if (@_) {
		$self->{'country'} = shift @_;
		$self->_update_value();
	}

	return $self->{'country'};
}

1;

package Data::v::Param;

use overload
	'""'  => \&as_string,
	'cmp' => \&Data::Header::Fields::Value::cmp,
;

sub new {
	my $class = shift;
	return bless {
		@_
	}, $class;
}

sub name {
	my $self   = shift;	

	$self->{'name'} = shift @_
		if (@_);

	return $self->{'name'};
}

sub value {
	my $self   = shift;	

	$self->{'value'} = shift @_
		if (@_);

	return $self->{'value'};
}

sub as_string {
	my $self   = shift;	
	
	return
		(lc $self->name eq 'type')
		? $self->value
		: $self->name.'='.$self->value
	;
}

1;

package Data::v::Calendar;

use base 'Data::v';

1;


__END__

=head1 SEE ALSO

L<http://tools.ietf.org/html/rfc2425> - A MIME Content-Type for Directory Information

L<http://tools.ietf.org/html/rfc2426> - vCard MIME Directory Profile

L<http://tools.ietf.org/html/rfc5545> - Internet Calendaring and Scheduling Core Object Specification (iCalendar)


=head1 AUTHOR

Jozef Kutej

=cut


lib/Data/Header/Fields.pm
package Data::Header::Fields;

use warnings;
use strict;

use IO::Any;
use Carp 'croak';
use String::Escape ();
use List::MoreUtils 'uniq', 'all';

use overload
	'""' => \&as_string,
	'cmp' => \&cmp,
;

our $VERSION = '0.04';

sub new {
	my $class = shift;
	return bless {
		tight_folding => 0,
		key_cmp       => sub { $_[0] cmp $_[1] },
		@_
	}, $class;
}

sub _lines {
	my $self = shift;
	
	$self->{_lines} = shift
		if (@_);
	
	$self->{_lines} = []
		if (not $self->{_lines});
	
	return $self->{_lines};
}

sub key_cmp {
	my $self   = shift;
	$self->{'key_cmp'} = shift
		if @_;
	
	return $self->{'key_cmp'};
}

sub tight_folding {
	my $self = shift;
	
	$self->{tight_folding} = shift
		if (@_);
	
	return 0
		if not ref $self;
	
	return $self->{tight_folding};
}

sub decode {
	my $self = shift;
	my $any  = shift;
	
	my @lines = $self->_read_lines($any);

	my $line_ending = (
		($lines[0] || '') =~ m/\r\n \Z/xms
		? "\r\n"
		: "\n"
	);
	$self->line_ending($line_ending);
	
	@lines = (
		map {
			Data::Header::Fields::Line->new(
				'line'          => $_,
				'parent'        => $self,
			);
		} @lines
	);
	
	if (ref $self) {
		$self->_lines(\@lines);
		return $self;
	}
	
	return \@lines;
}

sub _read_lines {
	my $self = shift;
	my $any  = shift;
	
	my $fh = IO::Any->read($any);
	
	# put folded lines to an array http://tools.ietf.org/html/rfc2822#section-2.2.3
	my @lines;
	while (my $line = <$fh>) {
		# folded line
		if (($line =~ m/^\s/xms)) {
			# ignore if the first line starts with white space
			next if not @lines;
			
			$lines[-1] .= $line;
			next;
		}
		push @lines, $line;
	}
	
	close $fh;

	return @lines;	
}

*as_string = *encode;
sub encode {
	my $self  = shift;
	my $lines = shift || (ref $self ? $self->_lines : undef);
	
	# no additional arguments
	if (@_ == 0) {
		my $text = '';
		$self->encode($lines, \$text);
		return $text;
	}
	
	my $any = shift;
	
	my $fh = IO::Any->write($any);
	foreach my $line (@{$lines}) {
		print $fh $line->as_string;
	}
	
	close $fh;
	
	return $self;
}

sub get_fields {
	my $self       = shift;
	my $field_name = shift or croak 'field_name argument is mandatory';
	
	my $key_cmp = $self->key_cmp;
	return (
		grep {
			$key_cmp->($field_name, $_->key) == 0
		} @{$self->_lines}
	);
}

sub get_field {
	my $self       = shift;
	my $field_name = shift or croak 'field_name argument is mandatory';
	my @extra_args = @_;
	
	my @fields = $self->get_fields($field_name, @extra_args);
	croak 'more then one header field with name "'.$field_name.'"'
		if @fields > 1;
	
	return $fields[0];
}

sub get_value {
	my $self = shift;
	my $key  = shift or croak 'key argument is mandatory';
	my @extra_args = @_;

	my $field = $self->get_field($key, @extra_args);
	return undef if not defined $field;
	return $field->value;
}

sub update_values {
	my $self  = shift;
	my $key   = shift or croak 'key argument is mandatory';
	my $value = shift;

	my $key_cmp = $self->key_cmp;
	my @lines = (
		map {
			($key_cmp->($_->key, $key) == 0 ? $_->value($value) : ());
			$_;
		} @{$self->_lines}
	);
	
	return $self;
}

sub rm_fields {
	my $self          = shift;
	my (@field_names) = (@_) or croak 'field_names argument is mandatory';

	my $key_cmp = $self->key_cmp;
	my @lines = (
		grep {
			my $key = $_->key;
			all { $key_cmp->($key, $_) != 0 } @field_names;
		} @{$self->_lines}
	);
	
	$self->_lines(\@lines);
	
	return $self;
}

sub set_value {
	my $self  = shift;
	my $key   = shift or croak 'key argument is mandatory';
	my $value = shift;

	my @fields = $self->get_fields($key);
	if (@fields == 1) {
		$self->update_values($key, $value);
	}
	elsif (@fields == 0) {
		push @{$self->_lines}, Data::Header::Fields::Line->new(
			'key' => $key,
			'value' => $value,
			'parent' => $self,
		);
	}
	else { 
		croak 'more then one header field with name "'.$key.'"';
	}
	
	
	return $self;
}

sub cmp {
	my $a = shift;
	my $b = shift;
	
	$a = $a->encode if ref $a and $a->can('encode');
	$b = $b->encode if ref $b and $b->can('encode');
	
	return $a cmp $b;
}

sub keys {
	my $self  = shift;	
	my $lines = shift || (ref $self ? $self->_lines : []);
	
	return
		uniq
		map {
			$_->key
		} @{$lines}
	;
}

sub line_ending {
	my $self = shift;
	
	return "\n"
		if not ref $self;
	
	if (@_) {
		$self->{line_ending} = shift;
	}
	$self->{line_ending} = "\n"
		if (not $self->{line_ending});
	
	return $self->{line_ending};
}


1;

package Data::Header::Fields::Value;

use Scalar::Util 'weaken', 'isweak';

use overload
	'""' => \&as_string,
	'cmp' => \&cmp,
;

sub new {
	my $class = shift;
	my $value = shift;
	
	if (@_ == 0) {
		if (not ref $value) {
			$value = { 'value' => $value };
		}
	}
	else {
		$value = { $value, @_ };
	}
	
	my $self = bless { 'parent' => $class->_default_parent, %{$value} }, $class;
	
	weaken($self->{'parent'})
		if (ref($self->{'parent'}) && !isweak($self->{'parent'}));
	
	return $self;
}

sub as_string {
	my $self   = shift;

	# remove folding
	my $line = $self->{value};
	if ($self->parent->parent->tight_folding) {
		$line =~ s/\n\s//xmsg;
	}
	else {
		$line =~ s/\n(\s)/$1/xmsg;
	}
	$line =~ s/\r?\n$//;
	$line = String::Escape::unprintable($line);
	
	return $line;
}

sub cmp {
	my $a = shift;
	my $b = shift;
	
	$a = $a->as_string if ref $a and $a->can('as_string');
	$b = $b->as_string if ref $b and $b->can('as_string');
	
	return $a cmp $b;
}

sub _default_parent {
	return 'Data::Header::Fields::Line';
}

sub parent {
	my $self   = shift;
	$self->{'parent'} = shift
		if @_;
	
	return (ref $self->{'parent'} ? $self->{'parent'} : $self->_default_parent);
}

sub value {
	my $self = shift;
	
	if (@_) {
		$self->{'value'} = shift;
		$self->parent->line_changed;
	}
	
	return $self->{'value'};
}
1;

package Data::Header::Fields::Line;

use Scalar::Util 'blessed', 'weaken', 'isweak';

use overload
	'""' => \&as_string,
	'cmp' => \&cmp,
;

sub new {
	my $class = shift;
	my $line  = shift;
	my @args  = @_;
	
	if (@args > 0) {
		$line = { $line, @args };
	}
	
	if (not ref $line) {
		$line = { 'line' => $line };
	}
	
	$line = { 'parent' => $class->_default_parent, %{$line} };
	
	if (exists $line->{'line'}) {
		# reblessing the line object
		if (blessed $line->{'line'}) {
			my $self = delete $line->{'line'};
			foreach my $key (keys %{$line}) {
				$self->{$key} = $line->{$key};
			}
			return bless $self, $class;			
		}
		else {
			my $line_string   = delete $line->{'line'};
			$line->{'original_line'} = $line_string;
			my ($key, $value) = split(/:/, $line_string, 2);
			$line->{'key'}    = $key;
			$line->{'value'}  = Data::Header::Fields::Value->new(
				'value'         => $value,
				'parent'        => $line,
			);
		}
	}
	
	weaken($line->{'parent'})
		if (ref($line->{'parent'}) && !isweak($line->{'parent'}));
	
	return bless $line, $class;
}

sub key {
	my $self   = shift;
	$self->{'key'} = shift
		if @_;
	
	return $self->{'key'};
}
sub value {
	my $self   = shift;
	$self->line_changed->{'value'} = shift
		if @_;
	
	return $self->{'value'};
}

sub line_changed {
	my $self = shift;
	delete $self->{'original_line'}
		if ref $self;
	return $self;
}

sub as_string {
	my $self = shift;
	
	if (exists $self->{'original_line'}) {
		my $original_line = $self->{'original_line'};
		
		# make sure the line has line_ending, even the original one could be created without using ->new()
		$original_line .= $self->parent->line_ending
			if $original_line !~ m/ \n \Z /xms;
		
		return $original_line;
	}

	my ($key, $value) = ($self->key, $self->value);
	$value = String::Escape::printable($value);

	my $line = join(':', $key, $value);
	
	$line .= $self->parent->line_ending
		if $line !~ m/\n$/xms;
	
	return $line;
}

sub cmp {
	my $a = shift;
	my $b = shift;
	
	$a = $a->as_string if ref $a and $a->can('as_string');
	$b = $b->as_string if ref $b and $b->can('as_string');
	
	return $a cmp $b;
}

sub _default_parent {
	return 'Data::Header::Fields';
}

sub parent {
	my $self   = shift;
	$self->{'parent'} = shift
		if @_;
	
	return (ref $self ? $self->{'parent'} : $self->_default_parent);
}


1;


__END__

=head1 NAME

Data::Header::Fields - encode and decode RFC822 header field lines

=head1 SYNOPSIS

	use IO::Any;
	my $email_msg = IO::Any->slurp([ 'path', 'to', 'email.eml' ]);
	my ($email_header, $email_body) = split(/^\s*$/m, $email_msg, 2);

	use Data::Header::Fields;
	my $dhf = Data::Header::Fields->new->decode(\$email_header);
	print 'From    - ', $dhf->get_value('From'), "\n";
	print 'Subject - ', $dhf->get_value('Subject'), "\n";
	print 'Date    - ', $dhf->get_value('Date'), "\n";
	print '--- cut ---', "\n";

	$dhf->set_value('To' => ' anyone@anywhere');
	$dhf->rm_fields('Received');
	
	print $dhf->encode();

=head1 WARNING

experimental, use on your own risk :-)

=head1 DESCRIPTION

RFC822 - Standard for ARPA Internet Text Messages (L<http://tools.ietf.org/html/rfc822#section-3.2>)
describes the format of header lines used in emails. The tricky part is
the line folding.

There are some "forks" of this standard. One of them is Debian RFC-2822-like
fields and the other is RFC2425 that defines the so called vCard format.
L<Data::Header::Fields> is generic enough to serve as a base class to parse
those as well.

One of the main goals of the module is to be able to edit the headers while
keeping the lines that were not changed untouched.

For the moment this is all documentation. After more tests with vCards and
using this module for the basic parsing in L<Parse::Deb::Control> it will
be stable enough.

Currently this distribution is highly radioactive!

=head1 SEE ALSO

L<http://tools.ietf.org/html/rfc2822> - Internet Message Format

=head1 AUTHOR

Jozef Kutej

=cut
